raulin0/backup-juggler
# feat: adds backup feature

## Description
This pull request introduces several enhancements to the Backup Juggler application. The changes include the addition of a mkdocs.yml configuration file for documentation generation, the addition of documentation for the Backup Juggler API modules, the inclusion of dependencies and configuration in the pyproject.toml file, the addition of unit tests for the backup functionality, the addition of fixtures for creating temporary paths during testing, the implementation of a ProgressViewer class for displaying the progress of the file copying operation, the implementation of the file backup functionality, and the addition of a Paths model class for representing source and destination paths in the application.

## Changes Made
- Added mkdocs.yml configuration file to enable documentation generation using MkDocs. The configuration includes settings for site name, repository URL, edit link URL, theme selection, directories to watch for changes, and plugins for macros and generating documentation from Python docstrings.
- Added documentation for Backup Juggler API modules to provide comprehensive information on the application's functionality.
- Added dependencies and configuration to pyproject.toml file, including dependencies such as tqdm and rich, dev dependencies for testing and code formatting, and doc dependencies for MkDocs documentation generation.
- Added unit tests for the backup functionality, covering different scenarios for file and directory copying, and handling non-existent source files.
- Added fixtures for creating temporary paths during testing, including fixtures for creating files, directories, and nested directory structures.
- Implemented a ProgressViewer class responsible for displaying the progress of the file copying operation using a progress bar. The class utilizes the tqdm library.
- Implemented file backup functionality, including the FileCopier class for managing the file copying process, the calculates_size function for calculating the total size of source directories or files, the backup function for initiating a single backup operation, and the parallel_backups function for executing multiple file backup operations in parallel.
- Added a Paths model class to represent source and destination paths in the application. The class provides methods for retrieving source and destination paths and calculating the total size of files to be copied.

## Motivation and Context
These changes enhance the Backup Juggler application by improving its documentation generation, adding unit tests for robustness, introducing a progress viewer for better user experience, implementing file backup functionality, and providing a model class for managing source and destination paths. These enhancements contribute to the overall functionality, maintainability, and usability of the application.

## How to Test
1. Verify that the mkdocs.yml configuration file is properly set up and documentation can be generated using MkDocs.
2. Review the added documentation for the Backup Juggler API modules and ensure it provides comprehensive information on the application's functionality.
3. Check that the dependencies and configuration in the pyproject.toml file are correctly specified.
4. Run the unit tests for the backup functionality and verify that they cover various scenarios and produce the expected results.
5. Use the provided fixtures for creating temporary paths during testing and ensure they generate the desired file and directory structures.
6. Test the progress viewer by initiating a file copying operation and verifying that the progress bar updates accurately.
7. Perform file backup operations using different source and destination paths, and verify that the files are copied correctly.
8. Execute multiple file backup operations in parallel and ensure they complete successfully.
9. Test error handling by providing non-existent source files or directories and confirm that appropriate exceptions are raised.

## Related Issue(s)
N/A

## Checklist
- [x] Documentation has been added or updated to reflect the new functionality.

Please review and merge this pull request at your earliest convenience. Any feedback or suggestions for improvement are welcome. Thank you!
